exe/cli/cmd/deal.go

```diff
@@ -0,0 +1,174 @@
+
+		addr := viper.GetString("address")
+		duration := viper.GetUint64("duration")
+		path := viper.GetString("file")
+		mp := viper.GetUint64("maxPrice")
+		ps := viper.GetUint64("pieceSize")
+		l := viper.GetInt("limit")
+		o := viper.GetInt("offset")
+
+		if addr == "" {
+			Fatal(errors.New("deal command needs a wallet address"))
+		}
+
+		if duration == 0 {
+			Fatal(errors.New("deal command duration should be > 0"))
+		}
+
+		if mp == 0 {
+			Fatal(errors.New("maxPrice must be > 0"))
+		}
+
+		if ps == 0 {
+			Fatal(errors.New("pieceSize must be > 0"))
+		}
+
+		file, err := os.Open(path)
+		checkErr(err)
+
+		q := ask.Query{
+			MaxPrice:  mp,
+			PieceSize: ps,
+			Limit:     l,
+			Offset:    o,
+		}
+
+		s := spin.New("%s Querying network for available storage asks...")
+		s.Start()
+		asks, err := fcClient.Deals.AvailableAsks(asksCtx, q)
+		s.Stop()
+		checkErr(err)
+
+		if len(asks) == 0 {
+			Message("No available asks, exiting")
+			return
+		}
+
+		choices := make([]string, len(asks))
+		for i, ask := range asks {
+			choices[i] = fmt.Sprintf("Price of %v with min piece size %v expiring %v", ask.Price, ask.MinPieceSize, ask.Expiry)
+		}
+
+		selectedAsks := []int{}
+		prompt := &survey.MultiSelect{
+			Message: "Select from available asks:",
+			Options: choices,
+		}
+		survey.AskOne(prompt, &selectedAsks, survey.WithValidator(survey.Required))
+		cmd.Println()
+
+		dealConfigs := make([]deals.DealConfig, len(selectedAsks))
+		for i, val := range selectedAsks {
+			ask := asks[val]
+			dealConfigs[i] = deals.DealConfig{
+				Miner:      ask.Miner,
+				EpochPrice: types.NewInt(ask.Price),
+			}
+		}
+
+		storeCtx, storeCancel := context.WithTimeout(context.Background(), cmdTimeout)
+		defer storeCancel()
+
+		s = spin.New("%s Initiating selected storage deals...")
+		s.Start()
+		succeeded, failed, err := fcClient.Deals.Store(storeCtx, addr, file, dealConfigs, duration)
+		s.Stop()
+		checkErr(err)
+
+		if len(failed) > 0 {
+			Message("Failed to initialize %v deals:", len(failed))
+			cmd.Println()
+			data := make([][]string, len(failed))
+			for i, dealConfig := range failed {
+				data[i] = []string{
+					dealConfig.Miner,
+					strconv.Itoa(int(dealConfig.EpochPrice.Int64())),
+				}
+			}
+			RenderTable(os.Stdout, []string{"miner", "price"}, data)
+			cmd.Println()
+			time.Sleep(time.Second * 2)
+		}
+
+		if len(succeeded) == 0 {
+			Message("Failed to initialize any deals, exiting")
+			return
+		}
+
+		Message("Watching progress of %v deals:", len(succeeded))
+		cmd.Println()
+
+		state := make(map[string]*client.WatchEvent, len(succeeded))
+		for _, cid := range succeeded {
+			state[cid.String()] = nil
+		}
+
+		writer := uilive.New()
+		writer.Start()
+
+		updateOutput(writer, state)
+
+		watchCtx, watchCancel := context.WithTimeout(context.Background(), cmdTimeout)
+		defer watchCancel()
+
+		ch, err := fcClient.Deals.Watch(watchCtx, succeeded)
+		checkErr(err)
+
+		for {
+			event, ok := <-ch
+			if ok == false {
```

Comment:
Nit: `!ok`.
Also, I wonder if this should also output some error message since isn't expected (being a timeout or an error)

Reply:
Cool, fixed the nit. `!ok` isn't necessarily and error, right? It happens when the channel is closed because no more data will be sent over it.

Reply:
Also, this is a channel of `client.WatchEvent`:
```
type WatchEvent struct {
	Deal deals.DealInfo
	Err  error
}
```
So we explicitly pass any error along as part of a channel event.

Reply:
In the case of an error coming through in a channel event, the error is displayed in the cli output next to the corresponding deal id.

Reply:
I think our explicit handling of any error in the event is good, going to resolve this conversation. Feel free to chime in again if you don't agree.

Reply:
> Cool, fixed the nit. `!ok` isn't necessarily and error, right? It happens when the channel is closed because no more data will be sent over it.

Oh ok.. if the items are that struct, makes sense.

Reply:
If that's the case you can use `for e := range `, so it terminates automatically on channel close.